gateway/src/p2p.rs

```diff
@@ -0,0 +1,237 @@
+use common::{
+    gm_request_response::{self, Behaviour as GMBehaviour, Event as GMEvent, Request, Response},
+    meta_exchange::{
+        self, Behaviour as MetaExchangeBehaviour, Event as MEEvent, Response as MEResponse, Role,
+    },
+};
+use futures::StreamExt;
+use libp2p::dns::Transport as DnsTransport;
+use libp2p::{
+    Multiaddr, PeerId,
+    core::{transport::Transport as _, upgrade},
+    identity,
+    noise::{Config as NoiseConfig, Error as NoiseError},
+    ping::{Behaviour as PingBehaviour, Config as PingConfig, Event as PingEvent},
+    swarm::{Config as SwarmConfig, NetworkBehaviour, Swarm, SwarmEvent},
+    tcp::{Config as TcpConfig, tokio::Transport as TcpTokioTransport},
+    yamux::Config as YamuxConfig,
+};
+use libp2p_request_response::{Message as RequestResponseMessage, ProtocolSupport};
+use log::{debug, error, info, warn};
+use std::{collections::HashSet, time::Duration};
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+
+#[derive(NetworkBehaviour)]
+#[behaviour(out_event = "GatewayEvent")]
+struct GatewayBehaviour {
+    ping: PingBehaviour,
+    request_response: GMBehaviour,
+    meta_exchange: MetaExchangeBehaviour,
+}
+pub enum GatewayEvent {
+    Ping(PingEvent),
+    RequestResponse(GMEvent),
+    MetaExchange(MEEvent),
+}
+
+impl From<PingEvent> for GatewayEvent {
+    fn from(e: PingEvent) -> Self {
+        GatewayEvent::Ping(e)
+    }
+}
+
+impl From<GMEvent> for GatewayEvent {
+    fn from(e: GMEvent) -> Self {
+        GatewayEvent::RequestResponse(e)
+    }
+}
+
+impl From<MEEvent> for GatewayEvent {
+    fn from(e: MEEvent) -> Self {
+        GatewayEvent::MetaExchange(e)
+    }
+}
+
+/// Internal structure that holds channels for inter-module communication
+struct Channels {
+    tx_request: UnboundedSender<Request>,
```

Comment:
Hmm, this is also duplicated somehow.

Isn't there a linter rule to fine these? Worthy of looking into for the future, imho.

Reply:
I did many experiments on how to make it cleaner. Replaced with AsyncInterface